File: android/app/src/main/java/com/mochi/keyboard/features/themes/ThemesViewModel.kt
package com.mochi.keyboard.features.themes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mochi.keyboard.data.ThemeCache
import com.mochi.keyboard.data.ThemeRepository
import com.mochi.keyboard.model.KeyboardTheme
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface ThemesUiState {
    data object Loading : ThemesUiState
    data class Data(val themes: List<KeyboardTheme>) : ThemesUiState
    data object Empty : ThemesUiState
    data class Error(val message: String) : ThemesUiState
}

class ThemesViewModel(private val themeRepository: ThemeRepository) : ViewModel() {

    private val _uiState = MutableStateFlow<ThemesUiState>(ThemesUiState.Loading)
    val uiState: StateFlow<ThemesUiState> = _uiState.asStateFlow()

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _uiState.value = ThemesUiState.Loading
            runCatching { themeRepository.getPublishedThemes(limit = 30) }
                .onSuccess { themes ->
                    ThemeCache.put(themes)
                    _uiState.value = if (themes.isEmpty()) ThemesUiState.Empty else ThemesUiState.Data(themes)
                }
                .onFailure { e -> _uiState.value = ThemesUiState.Error(e.message ?: "Couldn't load themes.") }
        }
    }
}

File: ios/MochiApp/Features/Themes/ThemesViewModel.swift
import Foundation

enum ThemesUiState {
    case loading
    case data([KeyboardTheme])
    case empty
    case error(String)
}

/// Swift mirror of android/.../features/themes/ThemesViewModel.kt — same Loading/Data/Empty/Error
/// shape and the same "any non-Data state falls back to MockData" convention ThemesScreen.kt uses
/// (unlike Home, which shows a real empty row for `.empty` — Themes always shows *something*).
@MainActor
final class ThemesViewModel: ObservableObject {
    @Published private(set) var uiState: ThemesUiState = .loading

    private let themeRepository: ThemeRepository?

    /// `nil` container (no GoogleService-Info.plist yet) means there's no real backend to query —
    /// same as every other screen, this degrades to permanently showing MockData.
    init(container: AppContainer?) {
        themeRepository = container?.themeRepository
        load()
    }

    func load() {
        guard let themeRepository else { return }
        uiState = .loading
        Task {
            do {
                let themes = try await themeRepository.getPublishedThemes(limit: 30)
                uiState = themes.isEmpty ? .empty : .data(themes)
            } catch {
                uiState = .error(error.localizedDescription)
            }
        }
    }
}
